mcstatus/status_response.py

```diff
@@ -0,0 +1,317 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+
+__all__ = [
+    "STYLE_MAP",
+    "AbstractDataclass",
+    "MCStatusResponse",
+    "JavaStatusResponse",
+    "BedrockStatusResponse",
+    "StatusPlayers",
+    "JavaStatusPlayers",
+    "BedrockStatusPlayers",
+    "JavaStatusPlayer",
+    "StatusVersion",
+    "JavaStatusVersion",
+    "BedrockStatusVersion",
+]
+
+STYLE_MAP = {
+    "bold": "l",
+    "italic": "o",
+    "underlined": "n",
+    "obfuscated": "k",
+    "color": {
+        "dark_red": "4",
+        "red": "c",
+        "gold": "6",
+        "yellow": "e",
+        "dark_green": "2",
+        "green": "a",
+        "aqua": "b",
+        "dark_aqua": "3",
+        "dark_blue": "1",
+        "blue": "9",
+        "light_purple": "d",
+        "dark_purple": "5",
+        "white": "f",
+        "gray": "7",
+        "dark_gray": "8",
+        "black": "0",
+    },
+}
+
+
+def _validate_data(raw: Dict[str, Any], who: str, required: Iterable[Tuple[str, type]]) -> None:
+    """Ensure that all required keys are present, and have the specified type.
+
+    :param raw: The raw dict answer to check.
+    :param who: The name of the object that is checking the data. Example `status`, `player` etc.
+    :param required: An iterable of string and type. The string is the required key which must be in `raw`, and
+        the `type` is the type that the key must be. If you want to ignore check of the type, set the type to `object`.
+    :raises ValueError: If the required keys are not present.
+    :raises TypeError: If the required keys are not of the specified type.
+    """
+    for required_key, required_type in required:
+        if required_key not in raw:
+            raise ValueError(f"Invalid {who} object (no '{required_key}' value)")
+        if not isinstance(raw[required_key], required_type):
+            raise TypeError(
+                f"Invalid {who} object (expected '{required_key}' to be {required_type}, was {type(raw[required_key])})"
+            )
+
+
+@dataclass
+class AbstractDataclass(ABC):
+    """Abstract base class for all dataclasses.
+
+    We can't use just `ABC` because we need raise TypeError if we try to
+    instantiate an abstract dataclass. Class which inherit from `ABC` raises
+    it only if it has abstract methods.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if cls is AbstractDataclass or cls.__bases__[0] is AbstractDataclass:
+            raise TypeError("Cannot instantiate abstract class.")
+        return super().__new__(cls)
```

Comment:
I doesn't see any sense in validation, `dict` object anyway will raise error on `my_dict["not_exist"]`.
Yes, it will less human-readable, but this exceptions (as I understand) not for humans.

This also check types, but types usually only for linters. Usually all ignore type-checking in runtime.

Reply:
Type checking in `PingResponse` was added by Dinnerbone in https://github.com/py-mine/mcstatus/commit/94264f22ea70cf8c07bbab6f97f647cde3b1cde2, so do we need it?

Also I using child classes (even without any new logic) for two reasons.
1. Someone, who will inspect the code in future, wouldn't ask a question like:
> Why are some classes called `StatusVersion` and others like `JavaStatusVersion`? How about adding individual classes to make it clearer?
2. Less-possible. Something will change in server answer, and you will need to implement this.

Reply:
I'm not saying we shouldn't be using child classes, I'm saying we don't need the special `AbstractDataclass`, because using simple `abc.ABC` works just fine with dataclasses. This comment was purely about removing this class, I'm not suggesting to remove the base classes we're using, I'm just suggesting to use simple `abc.ABC` with them, instead of this custom class.

All this custom class does is prevent direct initialization, no matter if the dataclass has some abstract methods or not. This same behavior is normal for `abc.ABC`, no matter if the class is a dataclass or not. It does make sense, because only classes with some methods without implementation shouldn't be possible to get initialized, classes without these methods, even if they're just base classes doesn't need to have initialization prevented completely.

I just suggested adding these checks into java response classes because they would mean an abstract `build` method would be present in the base classes, hence even with just simple `abc.ABC`, they wouldn't be initializable. That class can also just be a wrapper around `__init__` and just return `cls(*a, **kw)`. or maybe it doesn't need to be there at all and initialization can be allowed.

If you want to consider removing these type checks completely, make an issue about that, or do it in another PR, this one is already very big and doesn't need any further changes like this, that's a different discussion completely.

Reply:
> It does make sense, because only classes with some methods without implementation shouldn't be possible to get initialized, classes without these methods, even if they're just base classes doesn't need to have initialization prevented completely.

For dataclass this makes sense, because all what they should contain - just fields. And redef/add them should child classes.

Reply:
> ... This limitation should be sufficient since most of the abstract classes we try to define will contain the `build` method, which for non-concrete classes should stay abstract.
>
> Hm, although now that I'm looking at the rest of the code, it appears that `StatusPlayers` doesn't have this build abstractmethod, because it's concrete `BedrockStatusPlayers` subclass doesn't seem to need it.
>
> Though it could still be added to it, purely as a wrapper around `__init__`, or perhaps it could **include some value validation, similar to it's Java counterpart.**

I can't really imagine how I should add type checking there, because `decoded_data` which passing to `BedrockStatusResponse.build` full of strange numbers and strings. Also, it is all in `string` type because we decode it with `str.split` method.
![image](https://user-images.githubusercontent.com/68118654/168482921-a911e669-c4c5-4176-9139-542c07615a1a.png)
Of course I can go to tests and know what everything of they mean, but this will fail [later in `build` method](https://github.com/py-mine/mcstatus/pull/306/files#diff-1b85acf568c900bbfe553f0fdd02d0e8288b39a4c334a1c005e798d944c20762R214).

Reply:
> Of course I can go to tests and know what everyone of they mean, but types anyway this will fail [later in build method](https://github.com/py-mine/mcstatus/pull/306/files#diff-1b85acf568c900bbfe553f0fdd02d0e8288b39a4c334a1c005e798d944c20762R214).

Actually, I can't. There is only one test which calling `BedrockStatusResponse.build`, and I'm bad in googling some low-level info.

Reply:
https://wiki.
vg/ is a great resource for lookup up how minecraft protocol works on the low level, both for bedrock and for java. But yeah, adding type checking doesn't seem to make sense here, so let's either just add a an abstract method for build, and in bedrock it can simply act as a pure wrapper around init, so it essentially just returns an instance of that class:
```py
class Foo:
    def __init__(self, x):
        self.x = x

    @classmethod
    def build(cls, x):
        return cls(x)
```
or we could just allow initialization since there are no abstract methods, and it isn't a big issue anyway.

Reply:
Or we can just left `AbstractDataclass` as it is.

Reply:
Yes, but the whole point of this requested change was to not do that... `AbstractDataclass` doesn't have anything to do with dataclasses, the only thing it does is prevent initialization of it's child classes, and of itself, there is no need for this class since we don't need to block initialization of these dataclasses, even though they are only base classes.

`abc.ABC` does work with dataclasses just as well as it does with regular classes, it prevents initialization if those classes have some abstract methods defined. There is no reason why this behavior wouldn't be sufficient for these classes, and introducing `AbstractDataclass` makes the code much more complex without almost any benefit to it.

Reply:
> ... makes the code much more complex ...

Complex? It just preventing initialization of classes which never should be initialized. 

This class preventing initialization abstract classes, so in that way user can be sure that he got `JavaStatusResponse` or `BedrockStatusResponse`. Not an abstract class, without any real code.

Some classes doesn't need type check, because it useless. I also planning remove type checks in another PR, because they complitely useless. There is no reason to check types in runtime, moreover, parameters specified by `mcstatus` code, not by user code. And moreover, `__init__` will crash later, even without checks.

I don't want create stub methods, which never will be used. Why we need `build` if we already have `__init__`?

Reply:
There is absolutely no harm in letting those classes be initializable, and everything that doesn't need to be in the code, but is, is increasing complexity.

For people not that familiar with python, seeing `__new__` in a class can be very confusing and seem quite complex, and sure it technically isn't that complex if you're familiar with it and know how it works, but there are countless of people who just don't know that, and they shouldn't need to go through the struggle of looking it up just to understand this code, if it isn't even all that necessary. I don't mind having a bit more complex code in places where it's necessary, but if it can be avoid it, it should be avoided.

Type checking is not the issue here, it was just a thought I brought up as something that could fill the position of the `build` method, instead of it just being a wrapper. It is true that this method isn't necessary there though, I only suggested it if preventing initialization was that big of a concern, since it's a simpler way to do it than to have a class like this.

However again, there is no harm in letting these classes be initializable, the user shouldn't really even need to be initializing the full status response classes, since they're generated by the status methods on server classes. If they decide to initialize them, they probably know why they did that. There is no good reason to prevent it completely.

> This class preventing initialization abstract classes, so in that way user can be sure that he got JavaStatusResponse or BedrockStatusResponse. Not an abstract class, without any real code.

I don't see how a user could be confused on whether they got an base class or a real one, if they initialize it themselves, they will know what class is it, and if they get it from a method like `status`, it will certainly be the full class, not a base class.

Reply:
> For people not that familiar with python, seeing __new__ in a class can be very confusing and seem quite complex

Yes, it looks complex. But user (new in python here) will go to sources only if he want see interface, also abstract classes at all looks complex to new. This system, based on dataclasses and inheritance really hard to explain for new, because he need to know OOP and some others not so basic things.

Small ensurance that no one will initialize classes that mustn't be initialized not so big problem, and it is not so big harm delete it.

Also this code will inspected in only two cases:
1. Someone want to contribute and edit this code, so firstly he need to understand it all.
2. Someone get error when trying initialize abstract class. This can break something, so should be prevented.

Reply:
> Someone get error when trying initialize abstract class. This can break something, so should be prevented.

Initializing a base class can't break anything, the user won't even touch these classes unless he has explicitly searched for them, we don't and shouldn't be adding preventions like these because there is no reason to.

Keeping more code in the codebase, even though it can be removed is simply not a good idea, and there are a lot of people familiar with data-classes, and inheritance, and some even with abstract base classes, while not knowing about `__new__`, even some people who do have some experience with python won't know about `__new__`, because it's mostly not needed. It is quite rare that people would need to mess with it, and so we shouldn't expect most people to know about it.

> Someone want to contribute and edit this code, so firstly he need to understand it all.

If someone just wants to contribute a small fix to the codebase, they certainly shouldn't need to be familiar with every little aspect of it. While yes, if they want to understand the project and perhaps to some more work for it in the future going through the whole codebase is something that's worth doing, especially for a project that's not that big like mcstatus, but it shouldn't be a requirement.

And once again, since many python programmers don't know about `__new__`, some contributors could easily get discouraged to contribute because of the complexity of this code. Just because you know how to do something doesn't mean it should be done.

Python never tried too hard to prevent people from accessing private attributes, or preventing some behaviors. We shouldn't take it upon ourselves to limit some features that really don't need to be limited.

Reply:
I can add link to stackoverflow theme, where I copied it, so this class will be clear.

Reply:
i think it can simply be removed

Reply:
Idea of the `abc.ABC` is - not let initialize classes without implementation, but give general interface for all child classes.

`dataclass` package already build implementation methods without any interaction from us, so we need just write fields and some additional info to them.

So dataclass should just contain fields, and not enything else.

Some of our classes here have methods, and some are not. But abstract dataclass - actually class without any provided logic, that must be added in children class, so why we should allow initialization of some abstract dataclasses? They are just giving interface, and nothing else. Additional implementation must be added in children classes.

Reply:
> Idea of the abc.ABC is - not let initialize classes without implementation, but give general interface for all child classes.

Well, the whole point of ABC's not being initializable if they contain abstract methods is that other concrete functions in that ABC may be relying on the abstract function which lacks implementation, and even `__init__` may be one of those functions. For that reason, these classes simply block intialization if that's the case.

Class that's just providing general interface for all child classes is called a base class, and it doesn't necessarily need to be abstract, in which case there's nothing that may cause issues when initializing, such as having methods without implementation getting called.

> But abstract dataclass - actually class without any provided logic, that must be added in children class, so why we should allow initialization of some abstract dataclasses?

With our case, these classes have no chance of "not working properly", since they're only really designed to be containers for some attributes (that's also why they're dataclasses), and there's no reason to prevent initialization since there's no possibility of accidentally calling an unimplemented method, since there aren't any actual abstract methods, they may be base classes, but they're not necessarily abstract base classes.

To be fair, some of these classes do in fact have abstract methods, making them abstract base (data)classes, and those should therefore inherit from `abc.ABC`, and have their initialization prevented. But for those classes which don't have any of these abstract methods, we don't need to do that.

Instead, for non-abstract base classes, the fact that they aren't necessarily something that will be used for anything meaningful should be mentioned in their docstring, and it should be somewhat obvious from their name (such as making the name start with `Base`, indicating that this is a base class, which will be inherited by other classes, but won't necessarily be used directly)

> They are just giving interface, and nothing else. Additional implementation must be added in children classes.

Yes, you can argue that the user shouldn't ever need to be initializing these, but that doesn't mean we absolutely have to literally prohibit their initialization. In some languages, this would indeed be the expected behavior, however python isn't one of these languages. By default, python doesn't really even support encapsulation, there's no such thing as a `private` or `public` function/class (there are some tricks to achieve that, but you're clearly not meant to do it). There are some conventions, such as the use of `_` prefix, indicating that this should only be used internally, but that's all it is, a convention.

Making something unaccessible isn't something we usually do in python, and if it's going to be done, there should be a really good reason for that, which in this case, there really isn't. Even though the class wasn't necessarily meant to be used like that, there is nothing that would technically go wrong if you initialized it, and there's isn't really the need for something to go wrong by us preventing it (unlike with abstract base classes, where initialization and further usage may lead to calls to unimplemented abstract methods).

if the user decided to initialize that class, just to store the things which are shared across both some bedrock and java server, why should we prevent him? We won't use that class anywhere directly, but that doesn't mean it literally shouldn't be usable.

Additionally, on top of this manual encapsulation being unpythonic, it's also quite complex and it may discourage some contributors, or people just looking over the code trying to understand how it works, and what's available.

I simply don't think we need this, and in my opinion, it just introduces more necessary maintenance, possible bugs, unpythonic behavior, needless complexity, and needless preventions for the user without any real benefit. I think it should simply be removed.

Reply:
> There are some conventions, such as the use of `_` prefix, indicating that this should only be used internally, but that's all it is, a convention.

You can also mark it with double underscore, so it will looks like:
```py
class MyClass:
    def __my_method(self):
        return "My Method"

MyClass()._MyClass__my_method()
# -> My Method
```

---

But, okay. I will remove it.

Reply:
Removed in 6d949b1ded0fb20bddf02a8933f85c5ad7a6bb2d.

Reply:
> You can also mark it with double underscore

Indeed, however that functionality isn't here for encapsulation, but rather to prevent name collisions. For example if some class has `__x` vatiable in it, and class which overrides it also defined `__x`, it wouldn't be the same variable.

This is because python automatically coverts these double underscored parameters into `_ClassName__parametername`, so for example `__x` in `Foo` class would be accessible from the instance as: `Foo()._Foo__x`.

Many people confuse this behavior with encapsulation, since you can't directly do `Foo.__x`, but as you can see, the variable is in fact still accessible even in this case, it was just renamed to prevent naming collisions.

Reply:
I actually described this somewhere upper.